mlir/include/mlir/Dialect/Ptr/IR/PtrOps.td

```diff
@@ -17,6 +17,75 @@ include "mlir/Interfaces/SideEffectInterfaces.td"
 include "mlir/Interfaces/ViewLikeInterface.td"
 include "mlir/IR/OpAsmInterface.td"
 
+//===----------------------------------------------------------------------===//
+// FromPtrOp
+//===----------------------------------------------------------------------===//
+
+def Ptr_FromPtrOp : Pointer_Op<"from_ptr", [
+    Pure, OptionalTypesMatchWith<"metadata type", "result", "metadata",
+            "PtrMetadataType::get(cast<PtrLikeTypeInterface>($_self))">
+  ]> {
+  let summary = "Casts a `!ptr.ptr` value to a ptr-like value.";
+  let description = [{
+    The `from_ptr` operation casts a `ptr` value to a ptr-like object. It's
+    important to note that:
+    - The ptr-like object cannot be a `!ptr.ptr`.
+    - The memory-space of both the `ptr` and ptr-like object must match.
+    - The cast is side-effect free.
+
+    If the ptr-like object type has metadata, then the operation expects the
+    metadata as an argument or expects that the flag `trivial_metadata` is set.
+    If `trivial_metadata` is set, then it is assumed that the metadata can be
+    reconstructed statically from the pointer-like type.
```

Comment:
Why do we need the `trivial_metadata` flag instead of just assuming it when metadata isn't provided?

Reply:
Let's suppose that we have `%v0: memref<f32, 0>` in which `alignedPtr != allocatedPtr`.  Then in the following example `%v1 != %v0`:

```mlir
%p = to.ptr %v : memref<f32, 0> -> !ptr.ptr<0>
%v1 = from_ptr %p : !ptr.ptr<0> -> memref<f32, 0>
```

So in that case the cast sequence is a lossy conversion, which might be fine for the user and it should be possible to do. However, I think it makes for a less buggy experience to explicitly state that the user is ignoring the metadata.  

Currently, the above IR will generate a verification error saying that `from_ptr` requires metadata because the type says so.

Reply:
> owever, I think it makes for a less buggy experience to explicitly state that the user is ignoring the metadata.

For a programming language, I guess so, but for the IR that seems like redundant information here.

Reply:
It also tells the compiler that the folding is valid, see my comment in the folder of `FromPtrOp`.

Reply:
> It also tells the compiler that the folding is valid

I don't quite follow: if the absence of the flag implies trivial_metadata, then you can fold just as well.

Reply:
It's not an implementation impossibility of the folding logic, but rather an affirmation to the compiler that the user guarantees it's safe to do it, even when the cast is lossy like with `memref`.

My main point with this flag is to make it explicit that there's something potentially hazardous taking place. Otherwise an user could end up with something like:
```mlir
%v1 = from_ptr %p : !ptr.ptr<0> -> memref<f32, 0>
memref.dealloc %v1 : memref<f32, 0>
```
Which is potentially dangerous, and no warnings were raised.  Instead having:

```mlir
%v1 = from_ptr %p trivial_metadata : !ptr.ptr<0> -> memref<f32, 0>
memref.dealloc %v1 : memref<f32, 0>
```
At least places the burden on the user of the op.

Reply:
Again: this is an IR, not a programming language: people aren't writing textual IR by hand.
We don't need to have redundant information encoded on an operation.

Reply:
I have removed the `trivial_metadata` flag.